Allow autocomplete based on phone sync

nickvergessen opened the pull request:
- [x] Based on #26029 
- [x] Store phonebook sync matches in a database
- [x] Remove "known user" matches when the user changes their phone number
- [x] Remove any matches when the user is deleted
- [x] Add a config option to "Restrict username autocompletion to users with phone sync matches" in the sharing settings
    - [x] (needs some thinking how we can phrase it so it works with the existing options in a good way)
- [x] Respect that setting on the autocomplete endpoint.
- [x] Write integration tests to assure this doesn't break

nickvergessen committed:
Add a KnownUsers database with model

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Store when a user knows another user based on phone number

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Delete matches when a user changes their phone number

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Delete matches when the user is being deleted

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Add a service to find out if a user knows another user

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Add a config setting to restrict autocompletion to phonebook matches

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Restrict autocompletion also based on the phonebook known users

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Add integration tests for autocomplete/get (similar to sharees API)

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Add a hint that the settings are OR based

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Add a setting to restrict returning a full match unless in phonebook or same group

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Fix CS hopefully

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Change label also in the acceptance tests

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Also execute the new tests on drone

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Also clear the knownUser when changing via the settings endpoint

Signed-off-by: Joas Schilling <coding@schilljs.com>

ChristophWurst reviewed (approved):
I squinted but the code looks fine

PVince81 commented on lib/private/KnownUser/KnownUserMapper.php:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	/**
+	 * @param string $knownTo
```
add comment that this is the user id (and not contact id or whatever else) ? (also in various other places)
or rename to `$knownToUserId` ?

and maybe add a description to make it easier to distinguish between "knownTo" and "knownUser".

PVince81 commented on lib/private/KnownUser/KnownUserMapper.php:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	/**
+	 * @param string $knownTo
```
Returns all user ids that... the brain gymnastics hurts a bit, a description would be good here

PVince81 commented on lib/private/KnownUser/KnownUserService.php:
```diff
@@ -0,0 +1,62 @@
+
+namespace OC\KnownUser;
+
+class KnownUserService {
```
great idea to introduce this :+1:

nickvergessen committed:
Bump version to run migration

Signed-off-by: Joas Schilling <coding@schilljs.com>

PVince81 commented on lib/private/KnownUser/KnownUserService.php:
```diff
@@ -0,0 +1,62 @@
+		return $this->mapper->deleteKnownUser($knownUser);
+	}
+
+	public function storeIsKnownToUser(string $knownTo, string $knownUser): void {
```
again confused who is the user who is known to who.
as an API consumer I wouldn't know where to pass which user and would resort to trial and error

PVince81 reviewed (approved):
:+1: code looks fine with minor comments

and one thing that disturbed me most was the naming of $knownTo and $knownUser which are easy to mix up

nickvergessen committed:
Rename some parameters and methods to make the API more clear

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Clean up the logic of the contacts store a bit

Signed-off-by: Joas Schilling <coding@schilljs.com>

PVince81 commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -165,24 +165,16 @@ private function filterContacts(IUser $self,
 						if ($mailAddress === $filter) {
-							$filterUser = false;
-							break;
+							return true;
```
it feels wrong now to return directly here. if we said we don't want to filter, it seems the old code flow would continue into later if statements.

sorry if my former misunderstanding and review comment lead to this change :-/

PVince81 reviewed (approved):
:+1:

nickvergessen commented:
/backport to stable21

skjnldsv commented:
@nickvergessen Documentation missing?

nickvergessen commented:
How to upload your phone's addressbook? Or the admin setting?

skjnldsv commented:
> How to upload your phone's addressbook? Or the admin setting?

admin setting
https://github.com/nextcloud/server/blob/d9f85220038b2421870a36f9b6a860d6444a988a/apps/settings/lib/Settings/Admin/Sharing.php#L88-L94

nickvergessen commented:
Added the new checkbox in https://github.com/nextcloud/documentation/pull/11241 and updated the screenshot and labels that were brought in by other changes in the meantime
